### PR TITLE
Core lightning plugin to intercept HTLCs

### PR DESCRIPTION
**Pull request by justinmoon**
TLDR this PR turns `ln_gateway` into a core-lightning "plugin" so that it can intercept HTLCs using the [htlc_accepted](https://github.com/ElementsProject/lightning/blob/master/doc/PLUGINS.md#htlc_accepted) hook. In the future this will allow the gateway to buy the preimage from the federation and complete incoming lightning payments. Right now a preimage is hard-coded. I made a new file for this plugin since it's not really done yet ...

This PR also creates a new `UserClient::create_invoice(amount)` method to create a lightning invoice which will allow the gateway to receive payment and (eventually) figure out which preimage to buy from the federation.

Perhaps I should squash commits or break this up into a couple PRs? I'll leave some comments inline ...

### CODE REVIEW

File: ln-gateway/Cargo.toml [hunk 1]
```diff
@@ -26,3 +27,6 @@ tide = "0.16.0"
 tracing = "0.1.26"
 tracing-subscriber = { version = "0.3.1", features = [ "env-filter" ] }
 tokio = {version = "1.0", features = ["full"]}
+
+# FIXME: kept getting `Error: attempted to set a logger after the logging system was already initialized`
+log = "0.4.17"
```
Comment (justinmoon) [inline ln-gateway/Cargo.toml:32]: It seemed like tracing and the `cln_plugin` logging system were in conflict somehow and I couldn't figure out exactly how. This is kind of weird because core-lightning plugins use stdin / stdout for IO and that might have something to do with it ...

File: ln-gateway/src/bin/cln.rs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+#[tokio::main]
+async fn main() -> Result<(), Error> {
+    // FIXME: kept getting `Error: attempted to set a logger after the logging system was already initialized`
+    // tracing_subscriber::fmt()
+    //     .with_env_filter(
+    //         EnvFilter::try_from_default_env().unwrap_or_else(|_| EnvFilter::new("info")),
+    //     )
+    //     .init();
```
Comment (justinmoon) [inline ln-gateway/src/bin/cln.rs:88]: Just commented out the tracing code because we need to figure out how to get this working again and remove the `log` dependency I added for reasons mentioned above ...

File: ln-gateway/src/bin/cln.rs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            options::Value::String("default-dont-use".into()),
+            "minimint config directory",
+        ))
+        .hook("htlc_accepted", htlc_accepted_handler)
+        .start()
+        .await?
+    {
+        let workdir = match plugin.option("minimint-cfg").expect("minimint-cfg missing") {
+            options::Value::String(workdir) => {
+                // FIXME: cln_plugin doesn't yet support optional parameters
```
Comment (justinmoon) [inline ln-gateway/src/bin/cln.rs:102]: I'm going to add this to `cln_plugin`

File: mint-client/src/clients/user.rs [hunk 4]
```diff
@@ -379,6 +388,49 @@ impl UserClient {
...
+        let invoice = InvoiceBuilder::new(Currency::Regtest)
+            .amount_milli_satoshis(amount.milli_sat)
+            .description("description".into())
+            .payment_hash(payment_hash)
+            .payment_secret(payment_secret)
+            .current_timestamp()
+            .min_final_cltv_expiry(144)
+            .payee_pub_key(node_public_key)
+            .private_route(gateway_route_hint)
+            .build_signed(|hash| self.context.secp.sign_recoverable(hash, &node_secret_key))?;
```
Comment (justinmoon) [inline mint-client/src/clients/user.rs:432]: Some of these parameters I just copied from examples in the docs. Don't know exactly what all of them mean ...

File: mint-client/src/clients/user.rs [hunk 5]
```diff
@@ -62,23 +67,32 @@ impl UserClient {
...
+                .context
+                .borrow_with_module_config(|cfg| &cfg.client.wallet),
+            fee_consensus: self.context.config.client.fee_consensus.clone(), // TODO: remove or put into context
         }
     }
 
+    // FIXME: we probably want a "gateway client"
+    pub fn gateway_api(&self) -> String {
+        self.context.config.gateway.api.clone()
+    }
```
Comment (justinmoon) [inline mint-client/src/clients/user.rs:94]: This is required for the POST to `{gateway_host}/pay_invoice`. Putting HTTP calls in `UserClient` doesn't seem right. Maybe a `GatewayClient` that can be accessed by the `UserClient` in the same way the other clients are?

**Comment by justinmoon** [comment]
Re-reading #69 I think I still need to initialize the plugin to write RPC requests to stdout instead of the rpc socket ...

**Comment by elsirion** [comment]
Yeah, that will make the rpc path in the GW config obsolete since cln will tell you where its RPC socket is on initialization (iirc).

**Comment by justinmoon** [comment]
> Yeah, that will make the rpc path in the GW config obsolete since cln will tell you where its RPC socket is on initialization (iirc).

This is correct. Core-lightning sends this in the "configuration" section of the "init" message that is sent while the plugin is intializing. 

Problem is that `cln_plugin` doesn't save this while handling the "init" message. Made a [PR](https://github.com/ElementsProject/lightning/pull/5279) to fix this.

File: ln-gateway/src/bin/cln.rs [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+    let gateway = LnGateway::from_config(Box::new(db), cfg).await;
+
+    let state = State {
+        gateway: Arc::new(gateway),
+    };
+
+    let mut app = tide::with_state(state);
+    app.at("/pay_invoice").post(pay_invoice);
+    app.listen("27.0.0.1:8080").await?;
```
Comment (litch) [inline ln-gateway/src/bin/cln.rs:76]: Should this be 127.0.0.1?

Comment (justinmoon) [reply ln-gateway/src/bin/cln.rs:76]: Hmmm yes
